solrmonitor/solrmonitor.go

```diff
@@ -565,21 +564,30 @@ func (coll *collection) updateReplicaVersionAndState(newState *CollectionState,
 }
 
 func (coll *collection) startMonitoringReplicaStatus() {
-	coll.mu.Lock()
-	defer coll.mu.Unlock()
-
 	path := coll.parent.solrRoot + "/collections/" + coll.name + "/state.json"
 
 	// TODO: need to revisit coll.isWatched flag(if zk disconnects?). we need to create watch once only Scott?
-	if !coll.isWatched && coll.isPRSEnabled() {
+	if !coll.hasWatch() && coll.isPRSEnabled() {
```

Comment:
This breaks the lock's protection. The critical section is read, act, write. If you unlock around just the read and writes you might as well not have any locks.

Reply:
Agree, we might endup creating more than one watch on same path. But its very unlikely. if you have strong opinion on that then I will change. let me know about that!

Reply:
After further inspection of the code we only have one writing thread, so this is fine, and solves the reentrant lock issue the function would have without this change.